mypyc/irbuild/specialize.py

```diff
@@ -135,13 +141,47 @@ def translate_safe_generator_call(
                     + [builder.accept(arg) for arg in expr.args[1:]]),
                 builder.node_type(expr), expr.line, expr.arg_kinds, expr.arg_names)
         else:
+            if callee.fullname == "builtins.tuple":
+                val = tuple_from_generator_helper(builder, expr.args[0])
+                if val is not None:
+                    return val
             return builder.call_refexpr_with_args(
                 expr, callee,
                 ([translate_list_comprehension(builder, expr.args[0])]
                     + [builder.accept(arg) for arg in expr.args[1:]]))
     return None
 
 
+def tuple_from_generator_helper(builder: IRBuilder,
+                                gen: GeneratorExpr) -> Optional[Value]:
+
+    if len(gen.sequences) == 1 and len(gen.condlists[0]) == 0:
+        # Currently we only optimize for simplest generator expression
+        rtype = builder.node_type(gen.sequences[0])
+        if ((is_list_rprimitive(rtype) or is_tuple_rprimitive(rtype))
+                and is_int_rprimitive(builder.node_type(gen.indices[0]))):
```

Comment:
There is no need to restrict the index types to integers. I think that this check can be removed.

Reply:
Otherwise the case `testComprehensionShadowBinder` would fail.

Reply:
I suspect that this means that there is a problem elsewhere. The optimization should work for cases like this:

```py
def f(a: List[str]) -> Tuple[str, ...]:
    return tuple(x + 'a' for x in a)
```